pkg/module/registry_gcs.go

```diff
@@ -0,0 +1,141 @@
+package module
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"google.golang.org/api/iterator"
+	"io"
+	"os"
+
+	"cloud.google.com/go/storage"
+)
+
+// GCSRegistry is a Registry implementation backed by Google Cloud Storage.
+type GCSRegistry struct {
+	sc           *storage.Client
+	bucket       string
+	bucketPrefix string
+}
+
+func (s *GCSRegistry) GetModule(ctx context.Context, namespace, name, provider, version string) (Module, error) {
+	o := s.sc.Bucket(s.bucket).Object(moduleObjectKey(namespace, name, provider, version, s.bucketPrefix))
+	attrs, err := o.Attrs(ctx)
+	if err != nil {
+		return Module{}, errors.Wrap(ErrNotFound, err.Error())
+	}
+
+	return Module{
+		Namespace: namespace,
+		Name:      attrs.Name,
+		Provider:  provider,
+		Version:   version,
+		/* https://www.terraform.io/docs/internals/module-registry-protocol.html#sample-response-1
+		e.g. "gcs::https://www.googleapis.com/storage/v1/modules/foomodule.zip
+		*/
+		DownloadURL: s.generateDownloadURL(attrs.Bucket, attrs.Name),
+	}, nil
+}
+
+func (s *GCSRegistry) ListModuleVersions(ctx context.Context, namespace, name, provider string) ([]Module, error) {
+	var modules []Module
+	prefix := moduleObjectKeyBase(namespace, name, provider, s.bucketPrefix)
+
+	query := &storage.Query{
+		Prefix: prefix,
+		//Delimiter: "/",
```

Comment:
Can we remove this line or do we want this configurable?

Reply:
Removed it.